go/arrow/array/list.go

```diff
@@ -618,19 +631,661 @@ func (b *baseListBuilder) UnmarshalJSON(data []byte) error {
 	return b.Unmarshal(dec)
 }
 
+// ListView represents an immutable sequence of array values defined by an
+// offset into a child array and a length.
+type ListView struct {
+	array
+	values  arrow.Array
+	offsets []int32
+	sizes   []int32
+}
+
+var _ VarLenListLike = (*ListView)(nil)
+
+func NewListViewData(data arrow.ArrayData) *ListView {
+	a := &ListView{}
+	a.refCount = 1
+	a.setData(data.(*Data))
+	return a
+}
+
+func (a *ListView) ListValues() arrow.Array { return a.values }
+
+func (a *ListView) ValueStr(i int) string {
+	if !a.IsValid(i) {
+		return NullValueStr
+	}
+	return string(a.GetOneForMarshal(i).(json.RawMessage))
+}
+
+func (a *ListView) String() string {
+	o := new(strings.Builder)
+	o.WriteString("[")
+	for i := 0; i < a.Len(); i++ {
+		if i > 0 {
+			o.WriteString(" ")
+		}
+		if !a.IsValid(i) {
+			o.WriteString(NullValueStr)
+			continue
+		}
+		sub := a.newListValue(i)
+		fmt.Fprintf(o, "%v", sub)
+		sub.Release()
+	}
+	o.WriteString("]")
+	return o.String()
+}
+
+func (a *ListView) newListValue(i int) arrow.Array {
+	beg, end := a.ValueOffsets(i)
+	return NewSlice(a.values, beg, end)
+}
+
+func (a *ListView) setData(data *Data) {
+	a.array.setData(data)
+	offsets := data.buffers[1]
```

Comment:
Should we check `len(data.buffers) >= 3` and panic if it isn't large enough?

Reply:
With `debug.Assert`? Isn't `[i]` access already checked in Go at runtime?

Reply:
With `debug.Assert`? Isn't `[i]` access already checked in Go at runtime?

Reply:
`debug.Assert` is probably a good idea.

You're correct that `[i]` is already checked at runtime, but in this case you'd get a panic that wouldn't have much context or ease to debug. Adding the `debug.Assert` would be good since it would provide a more useful message

Reply:
I added a `debug.Assert` to all four list types' `setData` now.